discv4/kademlia.go

```diff
@@ -0,0 +1,133 @@
+package discv4
+
+import (
+	"container/list"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/indexsupply/lib/enr"
+)
+
+const (
+	maxBucketSize  = 16
+	addrByteSize   = 256 // size in bytes of the node ID
+	bucketsCount   = 20  // very rare we will ever encounter a node closer than log distance 20 away
+	minLogDistance = addrByteSize + 1 - bucketsCount
+)
+
+type kademliaTable struct {
+	mu       sync.Mutex // protects buckets
+	selfNode *enr.ENR
+	buckets  [bucketsCount]kBucket
+}
+
+type bucketEntry struct {
+	node     *enr.ENR
+	lastSeen time.Time
+}
+
+// kBucket stores an ordered list of nodes, from
+// most recently seen (head) to least recently seen (tail). The size
+// of the list is at most maxBucketSize.
+type kBucket struct {
+	lru        *list.List
+	entriesMap map[string]*list.Element
+}
+
+// nodes returns a slice of all the nodes (ENR) stored in this bucket.
+func (bucket *kBucket) nodes() []*enr.ENR {
+	var result []*enr.ENR
+	for element := bucket.lru.Front(); element != nil; element = element.Next() {
+		result = append(result, element.Value.(*bucketEntry).node)
+	}
+	return result
+}
+
+// Store inserts a node into this particular k-bucket. If the k-bucket is full,
+// then the least recently seen node is evicted.
+func (bucket *kBucket) store(node *enr.ENR) {
+	if el, ok := bucket.entriesMap[node.NodeAddrHex()]; ok {
+		// cache hit; update
+		el.Value.(*bucketEntry).lastSeen = time.Now()
+		el.Value.(*bucketEntry).node = node
+		bucket.lru.MoveToFront(el)
+		return
+	}
+
+	newEntry := bucket.lru.PushFront(&bucketEntry{node: node, lastSeen: time.Now()})
+	bucket.entriesMap[node.NodeAddrHex()] = newEntry
+
+	if bucket.lru.Len() > maxBucketSize {
+		// evict least recently seen
+		last := bucket.lru.Back()
+		bucket.lru.Remove(last)
+		delete(bucket.entriesMap, last.Value.(*bucketEntry).node.NodeAddrHex())
+	}
+}
+
+func newKademliaTable(selfNode *enr.ENR) *kademliaTable {
+	t := &kademliaTable{
+		selfNode: selfNode,
+		buckets:  [bucketsCount]kBucket{},
+	}
+	// init lists
```

Comment:
Supposedly lists are ready to be used without explicit initialization. https://pkg.go.dev/container/list#List